numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +323,72 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!PyArg_UnpackTuple(
+            positional_args, "implement_array_function", 5, 5,
+            &implementation, &public_api, &relevant_args, &args, &kwargs)) {
+        return NULL;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries.
+     */
+    PyObject *like_key = PyUnicode_FromString("like");
+    if (PyDict_CheckExact(kwargs) && PyDict_Contains(kwargs, like_key)) {
+        PyDict_DelItem(kwargs, like_key);
+    }
+    Py_DECREF(like_key);
+
+    return array_implement_array_function_internal(
+        implementation, public_api, relevant_args, args, kwargs);
+}
+
+
+/*
+ * Implements the __array_function__ protocol for a C function. Added as an extension
+ * to NEP-18 in an effort to bring NEP-35 to life with minimal dispatch overhead.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_c_array_function(
+    const char *function_name, PyObject *args, PyObject *kwargs)
+{
+    if (kwargs == NULL)
+        return NULL;
+
+    PyObject *relevant_args;
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries. If that key isn't present, return NULL and
+     * let originating call to continue.
+     */
+    PyObject *like_key = PyUnicode_FromString("like");
+    if (PyDict_CheckExact(kwargs) && PyDict_Contains(kwargs, like_key)) {
+        relevant_args = PyTuple_Pack(1, PyDict_GetItem(kwargs, like_key));
+        PyDict_DelItem(kwargs, like_key);
+    }
+    else {
+        return NULL;
+    }
+    Py_DECREF(like_key);
+
+    PyObject *numpy_module_name = PyUnicode_FromString("numpy");
+    PyObject *numpy_module = PyImport_Import(numpy_module_name);
+    PyObject *public_api = PyObject_GetAttrString(numpy_module, function_name);
+
```

Comment:
All of these calls can fail

Reply:
Yup, I didn't spend much time on error-checking yet as I wanted to get the discussion started, I'll do that soon. :)

Reply:
Done.